Hide command input field better

jacekkopecky opened the pull request:
Also removed the seemingly unnecessary editorContainer; this may
affect people’s personal styles if anybody has applied styles to that
container element.

The change in spec-helper makes sure that the new code’s expectations
are met when testing.

This is an alternative solution to PR #776.

jacekkopecky commented:
Tested to work with `atom` and `one` themes.

bronson commented:
Looks great, but I kinda miss the blue outline telling me it's waiting for a key.  :)  First merge this, then maybe someone can think of an unobtrusive way of making the modes more visible...?

jacekkopecky commented:
I'm thinking about that. Plus, I use a whole different set of cursors, with an empty rectangle cursor for inactive (which is triggered on `f` and such because the input line is the active editor then) which tends to be enough of an indication in the heat of coding.
My cursor styles:
```
@cursor: #29ff2f;
@cursorhi: #00ff00;

atom-text-editor.vim-mode {
  &.command-mode,
  &.operator-pending-mode,
  &.insert-mode
  { &::shadow, & {
      .cursors.blink-off .cursor, .cursors .cursor {
        border: 2px solid @cursor;
        background: none;
        z-index: -1;
        visibility: visible !important;
        opacity: 1;
      }

      .tile {
        background-color: rgba(0,0,0,0) !important;
      }
    }
  }

  &.visual-mode
  { &::shadow, & {
      .cursors.blink-off .cursor, .cursors .cursor {
        border: 2px solid @cursorhi;
        background: none;
        visibility: visible !important;
        opacity: .5;
      }
    }
  }

  &.is-focused {
    &.command-mode,
    { &::shadow, & {
        .cursors.blink-off .cursor, .cursors .cursor {
          border: none;
          background: @cursor;
        }
      }
    }

    &.operator-pending-mode
    { &::shadow, & {
        .cursors.blink-off .cursor, .cursors .cursor {
          border: none;
          border-bottom: 10px solid @cursor;
        }
      }
    }

    &.insert-mode,
    { &::shadow, & {
        .cursors.blink-off .cursor, .cursors .cursor {
          border: none;
          border-left: 3px solid @cursor;
        }
      }
    }

    &.visual-mode
    { &::shadow, & {
        .cursors.blink-off .cursor, .cursors .cursor,
         {
          border: none;
          border-left: 3px solid @cursorhi;
          opacity: .5;
        }
      }
    }

    &.replace-mode
    { &::shadow, & {
        .cursors.blink-off .cursor, .cursors .cursor {
          border: none;
          border-bottom: 4px solid @cursor;
        }
      }
    }
  }
}
```

bronson commented:
I see you use a light colorscheme!

I really like the rectangle operator-pending cursor.  Seems like it merits its own PR.  I couldn't manage to get it working with my dark colorscheme though, the cursor just disappears when going operator pending.  Gotta fiddle with it later.

bronson commented:
Just opened a speculative operator-pending PR in #779.

maxbrunsfeld commented:
I'd like to just *not* add the panel at all if the input is meant to be hidden. The problem is that the editor needs to be somewhere on the DOM in order to be focused. Maybe could just make it `1px x 1px` and add it to the body somewhere?

jacekkopecky commented:
@maxbrunsfeld this new and updated version does that - for a hidden input, the editor is invisible, added to the main text editor's parent element. Do you like this better?

maxbrunsfeld commented on lib/view-models/vim-normal-mode-input-element.coffee:
```diff
@@ -55,7 +51,10 @@ class VimNormalModeInputElement extends HTMLDivElement
+    if @panel?
+      @panel.destroy()
+    else
+      this.parentNode.removeChild(this) if this.parentNode?
```
I think you can avoid a conditional by using [`this.remove()`](https://developer.mozilla.org/en-US/docs/Web/API/ChildNode/remove).

jacekkopecky replied:
I keep forgetting about that one...

maxbrunsfeld commented on spec/operators-spec.coffee:
```diff
@@ -23,18 +23,19 @@ describe "Operators", ->
 
   describe "cancelling operations", ->
-    it "does not throw an error even if no operation is pending", ->
+    it "would throw an error when no operation is pending", ->
```
Thanks for catching this wrong description. I think it's better to simply say `it "throws an error` rather than `it "would throw an error`.

maxbrunsfeld commented:
Really glad to see this fixed. I left some very minor feedback.

jacekkopecky commented:
All done, thank you.

maxbrunsfeld commented:
:ok_hand: Great!
